src/contract.rs

```diff
@@ -242,11 +248,98 @@ pub fn query<S: Storage, A: Api, Q: Querier>(deps: &Extern<S, A, Q>, msg: QueryM
         QueryMsg::ContractStatus {} => query_contract_status(&deps.storage),
         QueryMsg::ExchangeRate {} => query_exchange_rate(&deps.storage),
         QueryMsg::Minters { .. } => query_minters(deps),
-        _ => authenticated_queries(deps, msg),
+        QueryMsg::WithPermit { permit, query } => query_with_permit(deps, permit, query),
+        _ => viewing_keys_queries(deps, msg),
     }
 }
 
-pub fn authenticated_queries<S: Storage, A: Api, Q: Querier>(
+fn query_with_permit<S: Storage, A: Api, Q: Querier>(
+    deps: &Extern<S, A, Q>,
+    permit: Permit,
+    query: QueryWithPermit,
+) -> Result<Binary, StdError> {
+    if permit.signed.msgs.len() != 1 {
+        return Err(StdError::generic_err(format!(
+            "Must sign exactly 1 permit message, got: {:?}.",
+            permit.signed.msgs.len()
+        )));
+    }
+
+    if permit.signed.msgs[0].r#type != "query_permit" {
+        return Err(StdError::generic_err(format!(
+            "Type must be 'query_permit', got: {:?}.",
+            permit.signed.msgs[0].r#type
+        )));
+    }
+
+    let permit_content = &permit.signed.msgs[0].value;
+
+    // Validate permit_name
+    let _permit_name = &permit_content.permit_name;
+    // TODO fail if permit_name is revoked
+
+    // Validate permit message
+    let token_address = ReadonlyConfig::from_storage(&deps.storage)
+        .constants()?
+        .contract_address;
+
+    if !permit_content.allowed_tokens.contains(&token_address) {
+        return Err(StdError::generic_err(format!(
+            "Permit doesn't apply to token {:?}, allowed tokens: {:?}",
+            token_address, permit_content.allowed_tokens
+        )));
+    }
+
```

Comment:
All the verification from here down can/should be a separate function in `permit.rs`